dev-docs/RFCs/v7.x-binary/glsl-accessor-rfc.md

```diff
@@ -0,0 +1,249 @@
+# RFC: GLSL Accessors
+
+* Authors: Ib Green
+* Date: June 13, 2018
+* Status: **Draft**
+
+This RFC is part of the [binary support roadmap](dev-docs/roadmaps/performance-roadmap.md).
+
+
+## Summary
+
+This RFC proposes supporting GLSL accessors (setting accessors to strings containing fragments of GLSL code), allowing layers to:
+
+* Completely or partially skip CPU side attribute generation
+* Work directly on supplied binary data (e.g. binary tables).
+* Do additional GPU processing on data, enabling "remarkably" performant custom effects.
+
+
+## Overview
+
+This RFC explores the idea of allowing modification of behavior of layers to be done on the GLSL level, not just the JS level.
+
+Let's start with a couple of API examples to illustrate what we want to achieve:
+
+
+### API Example: Using Columnar Binary Data Sirectly
+
+Columnar binary data systems (like Apache arrow) will often store primitive values (e.g. a Float32) in separate columns, rather than as a `vec2` or `vec3`. This means that even if we exposed these columns as attributes in GLSL, we cannot use them directly as the expected `vec3` attribute for positions. However, if we could just define a snippet of GLSL code to override how the shader extracs a position from attributes (`getPosition`), we could still use these values rather than pre-processing them in JavaScript:
+
+```
+new CircleLayer({
+
+  // inject new attributes
```

Comment:
Adding top-level, nested object props for customization like this can lead to a lot of issues. For all layer props we support one of the following path to update:

- update on shallow change - no, we definitely don't want to recompile the shader every frame
- update trigger - this is not an accessor
- update on deep change - it can still be expensive

Instead, I would like to roll shader injection into the layer extension proposal (extensions are wrapped in a class).